charger/delta.go

```diff
@@ -0,0 +1,318 @@
+	cc := struct {
+		Connector       uint16
+		modbus.Settings `mapstructure:",squash"`
+	}{
+		Connector: 1,
+		Settings: modbus.Settings{
+			ID: 1,
+		},
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewDelta(cc.URI, cc.Device, cc.Comset, cc.Baudrate, modbus.ProtocolFromRTU(cc.RTU), cc.ID, cc.Connector)
+}
+
+// NewDelta creates Delta charger
+func NewDelta(uri, device, comset string, baudrate int, proto modbus.Protocol, slaveID uint8, connector uint16) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, device, comset, baudrate, proto, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !sponsor.IsAuthorized() {
+		return nil, api.ErrSponsorRequired
+	}
+
+	log := util.NewLogger("delta")
+	conn.Logger(log.TRACE)
+
+	wb := &Delta{
+		log:  log,
+		conn: conn,
+		curr: 6000, // assume min current
+	}
+
+	wb.base = connector * 1000
+
+	// get failsafe timeout from charger
+	b, err := wb.conn.ReadHoldingRegisters(deltaRegCommunicationTimeout, 1)
+	if err != nil {
+		return nil, fmt.Errorf("failsafe timeout: %w", err)
+	}
+	if u := encoding.Uint16(b); u > 0 {
+		go wb.heartbeat(time.Duration(u) * time.Second / 2)
+	}
+
+	return wb, err
+}
+
+func (wb *Delta) heartbeat(timeout time.Duration) {
+	for range time.Tick(timeout) {
+		wb.mu.Lock()
+		var curr float64
+		if wb.enabled {
+			curr = wb.curr
+		}
+		wb.mu.Unlock()
+		if err := wb.setCurrent(curr); err != nil {
+			wb.log.ERROR.Println("heartbeat:", err)
+		}
+	}
+}
+
+// Status implements the api.Charger interface
+func (wb *Delta) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadInputRegisters(wb.base+deltaRegEvseChargerState, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	// 0: Charging process not started (no vehicle connected)
+	// 1: Connected, waiting for release (by RFID or local)
+	// 2: Charging process starts
+	// 3: Charging
+	// 4: Suspended (loading paused)
+	// 5: Charging process successfully completed (vehicle still plugged in)
+	// 6: Charging process completed by user (vehicle still plugged in)
+	// 7: Charging ended with error (vehicle still connected)
+	switch s := encoding.Uint16(b); s {
+	case 0:
+		return api.StatusA, nil
+	case 1, 2, 4, 5, 6, 7:
+		return api.StatusB, nil
+	case 3:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %0x", s)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Delta) Enabled() (bool, error) {
+	// b, err := wb.conn.ReadHoldingRegisters(wb.base+deltaRegEvseSuspendCharging, 1)
+	// if err != nil {
+	// 	return false, err
+	// }
+
+	// return encoding.Uint16(b) == 0, nil
+
+	b, err := wb.conn.ReadHoldingRegisters(wb.base+deltaRegEvseChargingPowerLimit, 2)
+	if err != nil {
+		return false, err
+	}
+
+	return encoding.Uint32(b) > 0, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *Delta) Enable(enable bool) error {
+	// b := make([]byte, 2)
```

Comment:
Kommentiertes bitte raus

Reply:
Das ist erstmal noch die Alternative und kann raus wenn klar ist dass die aktive Variante über Leistung 0 funktioniert.

Reply:
Sorry, irgendwie hat GH deinen Kommentar erst nach dem Refresh der Seite angezeigt. Habe den kommentierten Code rausgenommen. Wenn wir ihn doch noch brauchen sollten ist er ja in der History.